Sources/BSON/Date+BSONValue.swift

```diff
@@ -2,6 +2,50 @@ import Foundation
 import NIO
 
 extension Date: BSONValue {
+    /*
+     * Initializes a `Date` from ExtendedJSON.
+     *
+     * Parameters:
+     *   - `json`: a `JSON` representing the canonical or relaxed form of ExtendedJSON for a `Date`.
+     *   - `keyPath`: an array of `String`s containing the enclosing JSON keys of the current json being passed in.
+     *              This is used for error messages.
+     *
+     * Returns:
+     *   - `nil` if the provided value does not conform to the `Date` syntax.
+     *
+     * Throws:
+     *   - `DecodingError` if `json` is a partial match or is malformed.
+     */
+    internal init?(fromExtJSON json: JSON, keyPath: [String]) throws {
+        guard let (value, _) = try json.isObjectWithSingleKey(key: "$date", keyPath: keyPath) else {
```

Comment:
I know this was from the last review, but I just noticed now that we generally don't use the second part of this tuple, and in the places we do, I'm not sure we need to. For instance, in `BSONRegularExpression.swift`, we use it to print the keys in the error message, but the keys we actually want are those of the interior `regexObj`, not of `obj`. By nature of the method succeeding and returning non-nil, we know that `obj.keys` is `["$regularExpression"]` in that case. The same applies for `BSONBinary`.

Also, since this method is retrieving something, I think we should use a prefix that indicates that, since `is` is typically reserved for checking a condition and returning a boolean value.

Finally, to address the situations where this needs to be used with two different values (e.g. the interior object for `BSONRegularExpression`), we could use a similar method to reduce boilerplate. This doesn't scale well, but we only need it for exactly one or two values, so I think we can implement a method for this in an ad-hoc manner. Also, there is precedent for implementing methods like this in [Swift for TensorFlow](https://www.tensorflow.org/swift/api_docs/Protocols/Differentiable#sequencedthrough:_:) (see the various `sequenced` overloads).

To address these concerns, I suggest refactoring this method and also introducing a new one for two values:
```swift
extension JSON {
    func unwrapObject(withKeys: String, keyPath: [String]) throws -> (JSON)? {}
    func unwrapObject(withKeys key1: String, _ key2: String, keyPath: [String]) throws -> (JSON, JSON)? {}
}
```

Then usage would look like:
```swift
guard let (regexValue) = try json.unwrapObject(withKeys: "$regularExpression") else { ... }
guard let (pattern, options) = try regexValue.unwrapObject(withKeys: "pattern", "options") else { ... }
```